File: azure-cs-cosmosapp-component/CosmosApp.cs
// Copyright 2016-2018, Pulumi Corporation.  All rights reserved.

using System;
using System.Collections.Immutable;
using System.Linq;

using Pulumi;
using Pulumi.Azure.Core;
using Pulumi.Azure.CosmosDB;
using Pulumi.Azure.CosmosDB.Inputs;
using Pulumi.Azure.Network;
using Pulumi.Azure.Network.Inputs;

public class GlobalContext
{
    public ResourceGroup ResourceGroup { get; }
    public Account CosmosAccount { get; }
    public SqlDatabase Database { get; }
    public SqlContainer Container { get; }
    public CustomResourceOptions Options { get; }

    internal GlobalContext(ResourceGroup resourceGroup, Account cosmosAccount, SqlDatabase database, SqlContainer container, CustomResourceOptions options)
    {
        this.ResourceGroup = resourceGroup;
        this.CosmosAccount = cosmosAccount;
        this.Database = database;
        this.Container = container;
        this.Options = options;
    }
}

public class RegionalContext
{
    public string Location { get; }

    internal RegionalContext(string location)
    {
        this.Location = location;
    }
}

public interface IRegionalEndpoint
{
    // Type of the endpoint
    Input<string> Type { get; }
    // Azure resource ID (App Service and Public IP are supported)
    Input<string>? Id { get; }
    // An arbitrary URL for other resource types
    Input<string>? Url { get; }
}

public class AzureEndpoint : IRegionalEndpoint
{
    public Input<string> Type => "azureEndpoints";

    public Input<string> Id { get; }

    public Input<string>? Url => null;

    public AzureEndpoint(Input<string> id)
    {
        this.Id = id;
    }
}

public class ExternalEndpoint : IRegionalEndpoint
{
    public Input<string> Type => "externalEndpoints";

    public Input<string>? Id => null;

    public Input<string> Url { get;  }

    public ExternalEndpoint(Input<string> url)
    {
        this.Url = url;
    }
}

public class CosmosAppArgs
{
    public ResourceGroup ResourceGroup { get; set; } = null!;
    public string[] Locations { get; set; } = null!;
    public Func<GlobalContext, Func<RegionalContext, IRegionalEndpoint>> Factory { get; set; } = null!;
    public Input<string> DatabaseName { get; set; } = null!;
    public Input<string> ContainerName { get; set; } = null!;
    public Input<bool>? EnableMultiMaster { get; set; }
}

public class CosmosApp : ComponentResource
{
    public Output<string> Endpoint { get; private set; } = null!;

    public CosmosApp(string name, CosmosAppArgs args, ComponentResourceOptions? options = null)
        : base("examples:azure:CosmosApp", name, options)
    {
        var resourceGroup = args.ResourceGroup;
        var locations = args.Locations;
        var primaryLocation = locations[0];
        var parentOptions = new CustomResourceOptions { Parent = this };

        // Cosmos DB Account with multiple replicas
        var cosmosAccount = new Account($"cosmos-{name}",
            new AccountArgs
            {
                ResourceGroupName = resourceGroup.Name,
                Location = primaryLocation,
                GeoLocations = locations.Select((l, i) => new AccountGeoLocationArgs { Location = l, FailoverPriority = i }).ToArray(),
                OfferType = "Standard",
                ConsistencyPolicy = new AccountConsistencyPolicyArgs { ConsistencyLevel = "Session" },
                EnableMultipleWriteLocations = args.EnableMultiMaster,
            },
            parentOptions);

        var database = new SqlDatabase($"db-{name}",
            new SqlDatabaseArgs
            {
                ResourceGroupName = resourceGroup.Name,
                AccountName = cosmosAccount.Name,
                Name = args.DatabaseName,
            },
            parentOptions);

        var container = new SqlContainer($"sql-{name}",
            new SqlContainerArgs
            {
                ResourceGroupName = resourceGroup.Name,
                AccountName = cosmosAccount.Name,
                DatabaseName = database.Name,
                Name = args.ContainerName,
            },
            parentOptions);

        // Traffic Manager as a global HTTP endpoint
        var profile = new TrafficManagerProfile($"tm{name}",
            new TrafficManagerProfileArgs
            {
                ResourceGroupName = resourceGroup.Name,
                TrafficRoutingMethod = "Performance",
                DnsConfig = new TrafficManagerProfileDnsConfigArgs
                {
                    // Subdomain must be globally unique, so we default it with the full resource group name
                    RelativeName = Output.Format($"{name}{resourceGroup.Name}"),
                    Ttl = 60,
                },
                MonitorConfig = new TrafficManagerProfileMonitorConfigArgs
                {
                    Protocol = "HTTP",
                    Port = 80,
                    Path = "/api/ping",
                }
            },
            parentOptions);

        var globalContext = new GlobalContext(resourceGroup, cosmosAccount, database, container, parentOptions);
        var buildLocation = args.Factory(globalContext);
        var endpointOptions = new CustomResourceOptions { Parent = profile, DeleteBeforeReplace = true };

        var endpoints = locations.Select(location =>
        {
            var app = buildLocation(new RegionalContext(location));

            return new TrafficManagerEndpoint($"tm{name}{location}".Truncate(16),
                new TrafficManagerEndpointArgs
                {
                    ResourceGroupName = resourceGroup.Name,
                    ProfileName = profile.Name,
                    Type = app.Type,
                    TargetResourceId = app.Id,
                    Target = app.Url,
                    EndpointLocation = location,
                },
                endpointOptions);
        }).ToList();

        this.Endpoint = Output.Format($"http://{profile.Fqdn}");
        this.RegisterOutputs();
    }
}

File: classic-azure-ts-cosmosapp-component/cosmosApp.ts
// Copyright 2016-2018, Pulumi Corporation.  All rights reserved.

import * as azure from "@pulumi/azure";
import * as pulumi from "@pulumi/pulumi";

export interface GlobalContext {
    resourceGroup: azure.core.ResourceGroup;
    cosmosAccount: azure.cosmosdb.Account;
    database: azure.cosmosdb.SqlDatabase;
    container: azure.cosmosdb.SqlContainer;
    opts: pulumi.ResourceOptions;
}

export interface RegionalContext {
    location: string;
}

export interface RegionalEndpoint {
    // Azure resource ID (App Service and Public IP are supported)
    id?: pulumi.Input<string>;
    // An arbitrary URL for other resource types
    url?: pulumi.Input<string>;
}

type BuildLocation = (context: RegionalContext) => RegionalEndpoint;
type BuildLocationFactory = (context: GlobalContext) => BuildLocation;

export interface CosmosAppArgs {
    resourceGroup: azure.core.ResourceGroup;
    locations: pulumi.Input<pulumi.Input<string>[]>;
    databaseName: pulumi.Input<string>;
    containerName: pulumi.Input<string>;
    factory: BuildLocationFactory;
    enableMultiMaster?: boolean;
}

export class CosmosApp extends pulumi.ComponentResource {
    public endpoint: pulumi.Output<string>;

    constructor(name: string,
                args: CosmosAppArgs,
                opts: pulumi.ComponentResourceOptions = {}) {
        super("examples:azure:CosmosApp", name, args, opts);

        const resourceGroup = args.resourceGroup;
        const locations = pulumi.output(args.locations);
        const primaryLocation = locations[0];
        const parentOpts = { parent: this, ...opts };

        // Cosmos DB Account with multiple replicas
        const cosmosAccount = new azure.cosmosdb.Account(`cosmos-${name}`, {
            resourceGroupName: resourceGroup.name,
            location: primaryLocation,
            geoLocations: locations.apply(ls => ls.map((location, failoverPriority) => ({ location, failoverPriority }))),
            offerType: "Standard",
            consistencyPolicy: {
                consistencyLevel: "Session",
                maxIntervalInSeconds: 300,
                maxStalenessPrefix: 100000,
            },
            enableMultipleWriteLocations: args.enableMultiMaster,
        }, parentOpts);

        const database = new azure.cosmosdb.SqlDatabase(`db-${name}`, {
            resourceGroupName: resourceGroup.name,
            accountName: cosmosAccount.name,
            name: args.databaseName,
        }, { parent: cosmosAccount, ...opts });

        const container = new azure.cosmosdb.SqlContainer(`sql-${name}`, {
            resourceGroupName: resourceGroup.name,
            accountName: cosmosAccount.name,
            databaseName: database.name,
        }, { parent: cosmosAccount, ...opts });

        // Traffic Manager as a global HTTP endpoint
        const profile = new azure.network.TrafficManagerProfile(`tm${name}`, {
            resourceGroupName: resourceGroup.name,
            trafficRoutingMethod: "Performance",
            dnsConfig: {
                // Subdomain must be globally unique, so we default it with the full resource group name
                relativeName: pulumi.interpolate`${name}${resourceGroup.name}`,
                ttl: 60,
            },
            monitorConfig: {
                protocol: "HTTP",
                port: 80,
                path: "/api/ping",
            },
        }, parentOpts);

        const buildLocation = args.factory({ resourceGroup, cosmosAccount, database, container, opts: parentOpts });

        const endpoints = locations.apply(ls => ls.map(location => {
            const app = buildLocation({ location });

            // An endpoint per region for Traffic Manager, link to the corresponding instance
            return new azure.network.TrafficManagerEndpoint(`tm${name}${location}`.substring(0, 16), {
                resourceGroupName: resourceGroup.name,
                profileName: profile.name,
                type: app.id ? "azureEndpoints" : "externalEndpoints",
                targetResourceId: app.id,
                target: app.url,
                endpointLocation: location,
            }, { parent: profile, deleteBeforeReplace: true, ...opts });
        }));

        this.endpoint = endpoints.apply(_ => pulumi.interpolate`http://${profile.fqdn}`);

        this.registerOutputs();
    }
}
